oasislmf/exposures/csv_trans.py

```diff
@@ -6,65 +6,132 @@
+    multiprocess,
+    multithread,
+    Task,
+)
 
 class Translator(object):
-    def __init__(self, input_path, output_path, xsd_path, xslt_path, append_row_nums=False, chunk_size=5000, logger=None):
+    """
+    Transforms exposures/locations in CSV format
+    by converting a source file to XML and applying an XSLT transform
+    to apply rules for selecting, merging or updating columns
+
+    An optional step is to passing an XSD file for output validation
+
+    :param input_path: Source exposures file path, which should be in CSV comma delimited format 
+    :type input_path: str
+
+    :param output_path: File to write transform results
+    :type output_path: str
+
+    :param xslt_path: Source exposures Transformation rules file
+    :type xslt_path: str
+
+    :param xsd_path: Source exposures output validation file
+    :type xsd_path: str
+
+    :param append_row_nums: Append line numbers to first column of output called `ROW_ID` [1 .. n] when n is the number of rows processed. 
+    :type append_row_nums: boolean
+
+    :param chunk_size: Number of rows to process per multiprocess Task
+    :type chunk_size: int
+    """
+
+    def __init__(self, input_path, output_path, xslt_path, xsd_path=None, append_row_nums=False, chunk_size=5000, logger=None):
         self.logger = logger or logging.getLogger()
-        self.xsd = etree.parse(xsd_path)    # validation file
-        self.xslt = etree.parse(xslt_path)  # transform file
-        self.fpath_input = input_path       # file_in.csv
-        self.fpath_output = output_path     # file_out.csv
+        self.fpath_input = input_path
+        self.fpath_output = output_path
+        self.xslt = etree.parse(xslt_path)
+        self.xsd = (etree.parse(xsd_path) if xsd_path else None)
 
-        self.row_nums = append_row_nums  # Add 'ROW_ID' field to output
-        self.row_limit = chunk_size      # MAX number of CSV input rows
-        self.row_header_in = None        # CSV col header
-        self.row_header_out = None       # CSV col header Post Transform
+        self.row_nums = append_row_nums
+        self.row_limit = chunk_size
+        self.row_header_in = None
+        self.row_header_out = None
 
     def __call__(self):
         csv_reader = pd.read_csv(self.fpath_input, iterator=True, dtype=object, encoding='utf-8')
-        for data, first_row_number, last_row_number in self.next_file_slice(csv_reader):
-            self.logger.debug('--- lines[%d .. %d] ---', first_row_number, last_row_number)
 
-            # Convert CSV -> XML
-            xml_input_slice = self.csv_to_xml(
-                self.row_header_in,
-                data
+        task_list = []
+        for data, first_row, last_row in self.next_file_slice(csv_reader):
+            task_list.append(Task(self.process_chunk, (data,first_row,last_row)))
+        # run is process file chunk
+        for task in multithread(task_list):
+            pass
+
+        # write output
```

Comment:
If you assigned IDs to the chunks (enumerating `Translator.next_file_slice` will give you a pair `(i, it)` where `i` is a counter value and `it` is the next item generated) you might be able to order the chunks using the chunk ID. So you could also use `multiprocess` to create tasks for writing the chunks in the output.

This is a just a suggestion.